qiime2/sdk/result.py

```diff
@@ -556,3 +474,146 @@ def get_index_paths(self, relative=True):
+
+        if os.path.isfile(order_fp):
+            collection = cls._load_ordered(directory, order_fp)
+        else:
+            warnings.warn(f"The directory '{directory}' does not contain a "
+                          ".order file. The files will be read into the "
+                          "collection in the order the filesystem provides "
+                          "them in.")
+            collection = cls._load_unordered(directory)
+
+        return collection
+
+    @classmethod
+    def _load_ordered(cls, directory, order_fp):
+        collection = cls()
+
+        with open(order_fp, 'r') as order_fh:
+            for result_name in order_fh.read().splitlines():
+                result_fp = cls._get_result_fp(directory, result_name)
+                collection[result_name] = Result.load(result_fp)
+
+        return collection
+
+    @classmethod
+    def _load_unordered(cls, directory):
+        collection = cls()
+
+        for result in os.listdir(directory):
+            result_fp = os.path.join(directory, result)
+            result_name = result.rstrip('.qza')
+            result_name = result_name.rstrip('.qzv')
+
+            collection[result_name] = Result.load(result_fp)
+
+        return collection
+
+    @classmethod
+    def _get_result_fp(cls, directory, result_name):
+        result_fp = os.path.join(directory, result_name)
+
+        # Check if thing in .order file exists and if not try it with .qza at
+        # the end and if not try it with .qzv at the end
+        if not os.path.isfile(result_fp):
+            result_fp += '.qza'
+
+            if not os.path.isfile(result_fp):
+                # Get rid of the trailing .qza before adding .qzv
+                result_fp = result_fp[:-4]
+                result_fp += '.qzv'
+
+                if not os.path.isfile(result_fp):
+                    raise ValueError(
+                        f"The Result '{result_name}' is referenced in the "
+                        "order file but does not exist in the directory.")
+
+        return result_fp
+
+    def __init__(self, collection=None):
+        if collection is None:
+            self.collection = {}
+        elif isinstance(collection, dict):
+            self.collection = collection
+        else:
+            self.collection = {k: v for k, v in enumerate(collection)}
+
+    def __eq__(self, other):
+        if isinstance(other, dict):
+            return self.collection == other
+
+        return self.collection == other.collection
+
+    def __len__(self):
+        return len(self.collection)
+
+    def __iter__(self):
+        yield self.collection.__iter__()
+
+    def __setitem__(self, key, item):
+        self.collection[key] = item
+
+    def __getitem__(self, key):
+        return self.collection[key]
+
+    @property
+    def type(self):
+        # Get the type of the first element in the collection, all elements in
+        # the collection should have the same type.
```

Comment:
This is only true for _outputs_, inputs or user-constructed versions could have any type they want.

So long as there's some input which would accept `Collection[Foo | Bar]` that will work out fine. We just can't have that type as an output, as we would have to treat the union weirdly to compose correctly.

We can use `UnionExp(v.type for v in self.collection.values())` to create a really dumb looking Union, then we can call `.normalize()` on the result to reduce it to a normal form

Reply:
That is disgusting, but you're right